Windows Containerd support using Calico and HostProcess

jsturtevant opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature
<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind other
-->

**What this PR does / why we need it**:
Adds Containerd support for windows using Calico and hostprocess. This next step for capi windows proposal https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200804-windows-support.md#beta

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

requires 
- k8s 1.22+
- nightly build of containerd and hcsshim (https://github.com/kubernetes-sigs/sig-windows-tools/releases/tag/windows-containerd-nightly) 
- alpha Hostprocess features turned on
- images built in sig-windows https://github.com/kubernetes-sigs/sig-windows-tools/pull/161/commits

 currently `make tilt-up` works

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Windows Containerd support using Calico and HostProcess
```

/sig windows
/cc @marosset @perithompson

jsturtevant commented:
```
k get nodes -o wide
NAME                                          STATUS   ROLES                  AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION     CONTAINER-RUNTIME
win-conta-5zd7f                               Ready    <none>                 2m27s   v1.22.1   10.1.0.6      <none>        Windows Server 2019 Datacenter   10.0.17763.2114    containerd://1.5.0-beta.1-835-g79b369a0b
win-conta-vn9pc                               Ready    <none>                 2m37s   v1.22.1   10.1.0.4      <none>        Windows Server 2019 Datacenter   10.0.17763.2114    containerd://1.5.0-beta.1-835-g79b369a0b
win-containerd-template-control-plane-lmtbm   Ready    control-plane,master   4m32s   v1.22.1   10.0.0.4      <none>        Ubuntu 20.04.3 LTS               5.8.0-1039-azure   containerd://1.5.4
win-containerd-template-md-0-4cvnc            Ready    <none>                 3m9s    v1.22.1   10.1.0.5      <none>        Ubuntu 20.04.3 LTS               5.8.0-1039-azure   containerd://1.5.4
win-containerd-template-md-0-t2vtc            Ready    <none>                 3m13s   v1.22.1   10.1.0.7      <none>        Ubuntu 20.04.3 LTS               5.8.0-1039-azure   containerd://1.5.4
```

```
 k get pods -A      
NAMESPACE     NAME                                                                  READY   STATUS    RESTARTS      AGE
kube-system   calico-kube-controllers-846b5f484d-v86h4                              1/1     Running   0             4m18s
kube-system   calico-node-6t9bs                                                     1/1     Running   0             4m19s
kube-system   calico-node-l66tx                                                     1/1     Running   0             3m14s
kube-system   calico-node-vx5jb                                                     1/1     Running   0             3m18s
kube-system   calico-node-windows-fkcmk                                             2/2     Running   1 (31s ago)   2m32s
kube-system   calico-node-windows-g2p4v                                             2/2     Running   1 (43s ago)   2m42s
kube-system   coredns-78fcd69978-f7n8h                                              1/1     Running   0             4m28s
kube-system   coredns-78fcd69978-nm8kk                                              1/1     Running   0             4m28s
kube-system   etcd-win-containerd-template-control-plane-lmtbm                      1/1     Running   0             4m22s
kube-system   kube-apiserver-win-containerd-template-control-plane-lmtbm            1/1     Running   0             4m22s
kube-system   kube-controller-manager-win-containerd-template-control-plane-lmtbm   1/1     Running   0             4m22s
kube-system   kube-proxy-st8kc                                                      1/1     Running   0             4m28s
kube-system   kube-proxy-tvppk                                                      1/1     Running   0             3m18s
kube-system   kube-proxy-windows-5l4mb                                              1/1     Running   0             2m42s
kube-system   kube-proxy-windows-dd9j6                                              1/1     Running   0             2m32s
kube-system   kube-proxy-x7rrs                                                      1/1     Running   0             3m14s
kube-system   kube-scheduler-win-containerd-template-control-plane-lmtbm            1/1     Running   0             4m22s
```

perithompson commented on templates/addons/windows/calico/calico.yaml:
```diff
@@ -0,0 +1,238 @@
+        - name: VXLAN_VNI
+          value: "4096"
+        - name: KUBECONFIG
+          value: "C:/etc/cni/net.d/calico-kubeconfig"
```
could this reference a config map to save having to save the secret on the c: outside of the container root path?  I've been doing something similar, but it does need copying out to root for kube-proxy as that seems to be hardcoded to look in c:/var/lib/kube-proxy

jsturtevant replied:
This config file is for the calico.exe cni binary, which requires access to the api server and is invoked by CRI (via containerd) so I believe there is no way to avoid having this file.   It is re-used be the Felix service but we could potentially use the service account felix is linked too (but it has the same RBAC as the kubeconfig is generated by using the same SA)

perithompson replied:
yeah, I think I have found this too especially with kube-proxy, I guess I was just hoping that the container could write this file itself from a secret rather than have to provide it manually? unless it already is and I'm missing something?

jayunit100 replied:
id say merge and iterate a pluggable file path

perithompson commented on templates/addons/windows/calico/kube-proxy-windows.yaml:
```diff
@@ -0,0 +1,54 @@
+          runAsUserName: "NT AUTHORITY\\system"
+      hostNetwork: true
+      containers:
+      - image: sigwindowstools/kube-proxy:${KUBERNETES_VERSION/+/_}-calico-hostprocess
```
is this a special kube-proxy image?

jsturtevant replied:
Yes, it re-uses the scripts provided by kubeproxy (modified for this hostprocess) see more details at https://github.com/kubernetes-sigs/sig-windows-tools/pull/161/

jsturtevant replied:
eventually I would like to get the kubeproxy automatically deployed by kubeadm as well as make it configurable through kubeadm

perithompson replied:
interesting! thanks for the link to the other PR

jsturtevant commented:
Updated basic test include windows nodes now that we can easily support Calico. This adds two additional nodes instead of an entire cluster.  The unattended side affect is the `pull-cluster-api-provider-azure-e2e` didn't include that job any more (because of the [Windows exclusion in the job def](https://github.com/kubernetes/test-infra/blob/1279b0666472f59320b90365b583670ee3131c6b/config/jobs/kubernetes-sigs/cluster-api-provider-azure/cluster-api-provider-azure-presubmits.yaml#L66-L67))

but the job ran on the Windows Job and passed :tada:  

```
 [SLOW TEST:2984.834 seconds]
Workload cluster creation
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_test.go:43
  With 3 control-plane nodes and 2 worker nodes and 2 Windows worker nodes
  /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_test.go:203
------------------------------
STEP: Tearing down the management cluster
Ran 3 of 22 Specs in 3102.410 seconds 
```

I've updated the orginal Windows tests specs to include `dockershim` in the title. I would propose we exclude on `Dockershim + Windows` and rename the job Windows job to `Windows dockershim`. Once k8s 1.24 comes out we can remove the dockershim tests as it will no longer be in tree.

thoughts @CecileRobertMichon @devigned @marosset?

I am adding support for vmss and CI test templates now.

jsturtevant commented:
> I am adding support for vmss and CI test templates now.

Looking through the PR, it might make sense to add they in Separate PR to keep the size of this down.  Thoughts?

jsturtevant commented:
Ran into some issues with the e2e tests failing as noted in kubernetes/kubernetes#98102

I've been able to temporarily mitigate the issues and was able to get a pretty clean run:

```
Summarizing 3 Failures:

[Fail] [sig-windows] Services [It] should be able to create a functioning NodePort service for Windows 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/windows/hybrid_network.go:98

[Fail] [sig-network] Services [It] should be able to change the type from ExternalName to NodePort [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/service.go:1375

[Fail] [sig-network] Services [It] should be able to create a functioning NodePort service [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/service.go:1190

Ran 291 of 6853 Specs in 3226.564 seconds
FAIL! -- 288 Passed | 3 Failed | 0 Pending | 6562 Skipped
```

There is something wrong with node ports but the e2e test flow in CAPZ is working. I will begin looking into that but I don't know that it should block this PR.

jsturtevant commented:
/retest

jsturtevant commented:
/test ls

jsturtevant commented:
/test pull-cluster-api-provider-azure-conformance-with-ci-artifacts

jsturtevant commented:
/test pull-cluster-api-provider-azure-e2e-exp

jsturtevant commented:
The exp job looks like the flake is #1720

/test pull-cluster-api-provider-azure-e2e-exp
